// @angular-generic-table/core/esm5/components/gt-custom-component-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, EventEmitter, Injector, Input, Output, Type } from '@angular/core';
import { ReplaySubject } from 'rxjs/ReplaySubject';
/**
 * @abstract
 * @template R
 */
var /**
 * @abstract
 * @template R
 */
GtCustomComponent = /** @class */ (function () {
    function GtCustomComponent() {
        this.redrawEvent = new EventEmitter();
        this.$searchTerms = new ReplaySubject(1);
    }
    /**
     * @param {?=} $event
     * @return {?}
     */
    GtCustomComponent.prototype.$redraw = /**
     * @param {?=} $event
     * @return {?}
     */
    function ($event) {
        this.redrawEvent.emit({ row: this.row, column: this.column });
    };
    /**
     * @return {?}
     */
    GtCustomComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.$searchTerms.subscribe(function (terms) { return (_this.searchTerms = terms); });
    };
    return GtCustomComponent;
}());
/**
 * @abstract
 * @template R
 */
export { GtCustomComponent };
function GtCustomComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GtCustomComponent.prototype.row;
    /** @type {?} */
    GtCustomComponent.prototype.column;
    /** @type {?} */
    GtCustomComponent.prototype.redrawEvent;
    /** @type {?} */
    GtCustomComponent.prototype.searchTerms;
    /** @type {?} */
    GtCustomComponent.prototype.$searchTerms;
}
// unsupported: template constraints.
/**
 * @template R, C
 */
var GtCustomComponentFactory = /** @class */ (function () {
    function GtCustomComponentFactory() {
        this.$searchTerms = new ReplaySubject(1);
        this.redrawEvent = new EventEmitter();
    }
    Object.defineProperty(GtCustomComponentFactory.prototype, "searchTerms", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.$searchTerms.next(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} instance
     * @return {?}
     */
    GtCustomComponentFactory.prototype.instance = /**
     * @param {?} instance
     * @return {?}
     */
    function (instance) {
        instance.row = this.row;
        instance.column = this.column;
        instance.$searchTerms = this.$searchTerms;
        instance.redrawEvent.subscribe(this.redrawEvent);
    };
    GtCustomComponentFactory.decorators = [
        { type: Component, args: [{
                    selector: 'gt-custom-component-factory',
                    template: "<div gtComponentAnchor [ctor]=\"type\"\n                    [injector]=\"injector\"\n                    (instance)=\"instance($event)\"></div>"
                },] },
    ];
    /** @nocollapse */
    GtCustomComponentFactory.propDecorators = {
        "searchTerms": [{ type: Input },],
        "type": [{ type: Input },],
        "injector": [{ type: Input },],
        "row": [{ type: Input },],
        "column": [{ type: Input },],
        "redrawEvent": [{ type: Output },],
    };
    return GtCustomComponentFactory;
}());
export { GtCustomComponentFactory };
function GtCustomComponentFactory_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GtCustomComponentFactory.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GtCustomComponentFactory.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GtCustomComponentFactory.propDecorators;
    /** @type {?} */
    GtCustomComponentFactory.prototype.type;
    /** @type {?} */
    GtCustomComponentFactory.prototype.injector;
    /** @type {?} */
    GtCustomComponentFactory.prototype.row;
    /** @type {?} */
    GtCustomComponentFactory.prototype.column;
    /** @type {?} */
    GtCustomComponentFactory.prototype.$searchTerms;
    /** @type {?} */
    GtCustomComponentFactory.prototype.redrawEvent;
}
