// @angular-generic-table/core/esm5/components/gt-expanding-row.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, EventEmitter, Input, Output, Type } from '@angular/core';
// unsupported: template constraints.
/**
 * @template R
 */
var 
// unsupported: template constraints.
/**
 * @template R
 */
GtExpandedRow = /** @class */ (function () {
    function GtExpandedRow() {
        this.redrawEvent = new EventEmitter();
        this.toggleRowEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    GtExpandedRow.prototype.$hide = /**
     * @return {?}
     */
    function () {
        this.toggleRowEvent.emit(this.row);
    };
    /**
     * @param {?=} $event
     * @return {?}
     */
    GtExpandedRow.prototype.$redraw = /**
     * @param {?=} $event
     * @return {?}
     */
    function ($event) {
        this.redrawEvent.emit(this.row);
    };
    /**
     * @param {?} row
     * @param {?} $event
     * @return {?}
     */
    GtExpandedRow.prototype.$rowClick = /**
     * @param {?} row
     * @param {?} $event
     * @return {?}
     */
    function (row, $event) {
        this.gtEvent.emit({
            name: 'gt-row-clicked',
            value: { row: row, event: $event }
        });
    };
    return GtExpandedRow;
}());
// unsupported: template constraints.
/**
 * @template R
 */
export { GtExpandedRow };
function GtExpandedRow_tsickle_Closure_declarations() {
    /** @type {?} */
    GtExpandedRow.prototype.row;
    /** @type {?} */
    GtExpandedRow.prototype.columnWidth;
    /** @type {?} */
    GtExpandedRow.prototype.gtSettings;
    /** @type {?} */
    GtExpandedRow.prototype.gtFields;
    /** @type {?} */
    GtExpandedRow.prototype.gtOptions;
    /** @type {?} */
    GtExpandedRow.prototype.gtInfo;
    /** @type {?} */
    GtExpandedRow.prototype.data;
    /** @type {?} */
    GtExpandedRow.prototype.redrawEvent;
    /** @type {?} */
    GtExpandedRow.prototype.gtEvent;
    /** @type {?} */
    GtExpandedRow.prototype.toggleRowEvent;
}
// unsupported: template constraints.
// unsupported: template constraints.
/**
 * @template R, C
 */
var GtExpandingRowComponent = /** @class */ (function () {
    function GtExpandingRowComponent() {
        this.redrawEvent = new EventEmitter();
        this.toggleRowEvent = new EventEmitter();
    }
    /**
     * @param {?} instance
     * @return {?}
     */
    GtExpandingRowComponent.prototype.newInstance = /**
     * @param {?} instance
     * @return {?}
     */
    function (instance) {
        instance.row = this.row;
        instance.columnWidth = this.columnWidth;
        instance.gtSettings = this.gtSettings;
        instance.gtFields = this.gtFields;
        instance.gtOptions = this.gtOptions;
        instance.gtInfo = this.gtInfo;
        instance.data =
            typeof this.data === 'function' ? this.data(this.row) : this.data;
        instance.redrawEvent.subscribe(this.redrawEvent);
        instance.toggleRowEvent.subscribe(this.toggleRowEvent);
        instance.gtEvent = this.gtEvent;
    };
    GtExpandingRowComponent.decorators = [
        { type: Component, args: [{
                    selector: 'gt-expanding-row',
                    template: "\n        <div gtComponentAnchor\n             [ctor]=\"type\" (instance)=\"newInstance($event)\"></div>"
                },] },
    ];
    /** @nocollapse */
    GtExpandingRowComponent.propDecorators = {
        "type": [{ type: Input },],
        "row": [{ type: Input },],
        "columnWidth": [{ type: Input },],
        "gtSettings": [{ type: Input },],
        "gtFields": [{ type: Input },],
        "gtOptions": [{ type: Input },],
        "gtInfo": [{ type: Input },],
        "data": [{ type: Input },],
        "redrawEvent": [{ type: Output },],
        "toggleRowEvent": [{ type: Output },],
        "gtEvent": [{ type: Input },],
    };
    return GtExpandingRowComponent;
}());
export { GtExpandingRowComponent };
function GtExpandingRowComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GtExpandingRowComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GtExpandingRowComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GtExpandingRowComponent.propDecorators;
    /** @type {?} */
    GtExpandingRowComponent.prototype.type;
    /** @type {?} */
    GtExpandingRowComponent.prototype.row;
    /** @type {?} */
    GtExpandingRowComponent.prototype.columnWidth;
    /** @type {?} */
    GtExpandingRowComponent.prototype.gtSettings;
    /** @type {?} */
    GtExpandingRowComponent.prototype.gtFields;
    /** @type {?} */
    GtExpandingRowComponent.prototype.gtOptions;
    /** @type {?} */
    GtExpandingRowComponent.prototype.gtInfo;
    /** @type {?} */
    GtExpandingRowComponent.prototype.data;
    /** @type {?} */
    GtExpandingRowComponent.prototype.redrawEvent;
    /** @type {?} */
    GtExpandingRowComponent.prototype.toggleRowEvent;
    /** @type {?} */
    GtExpandingRowComponent.prototype.gtEvent;
}
